rcl/include/rcl/node.h

```diff
@@ -0,0 +1,231 @@
+ *         RCL_RET_ERROR if an unspecified error occurs.
+ */
+RCL_PUBLIC
+rcl_ret_t
+rcl_node_fini(rcl_node_t * node);
+
+/// Return the default node options in a rcl_node_options_t.
+RCL_PUBLIC
+rcl_node_options_t
+rcl_node_get_default_options();
+
+/// Get the name of the node.
+/* This function returns the node's internal name string.
+ * This function can fail, and therefore return NULL, if:
+ *   - node is NULL
+ *   - node has not been initialized (the implementation is invalid)
+ *
+ * The returned string is only valid as long as the given rcl_node_t is valid.
+ * The value of the string may change if the value in the rcl_node_t changes,
+ * and therefore copying the string is recommended if this is a concern.
+ *
+ * \param[in] node pointer to the node
+ * \return name string if successful, otherwise NULL
+ */
+RCL_PUBLIC
+const char *
+rcl_node_get_name(const rcl_node_t * node);
+
+/// Return the rcl node options.
+/* This function returns the node's internal options struct.
+ * This function can fail, and therefore return NULL, if:
+ *   - node is NULL
+ *   - node has not been initialized (the implementation is invalid)
+ *
+ * The returned struct is only valid as long as the given rcl_node_t is valid.
+ * The values in the struct may change if the options of the rcl_node_t changes,
+ * and therefore copying the struct is recommended if this is a concern.
+ *
+ * \param[in] node pointer to the node
+ * \return options struct if successful, otherwise NULL
+ */
+RCL_PUBLIC
+const rcl_node_options_t *
+rcl_node_get_options(const rcl_node_t * node);
+
+/// Return the ROS domain ID that the node is using.
+/* This function returns the ROS domain ID that the node is in.
+ *
+ * This function should be used to determine what domain_id was used rather
+ * than checking the domin_id field in the node options, because if
+ * RCL_NODE_OPTIONS_DEFAULT_DOMAIN_ID is used when creating the node then
+ * it is not changed after creation, but this function will return the actual
+ * domain_id used.
+ *
+ * The domain_id field must point to an allocated size_t object to which the
+ * ROS domain ID will be written.
+ *
+ * \param[in] node the handle to the node being queried
+ * \return RCL_RET_OK if node the domain ID was retrieved successfully, or
+ *         RCL_RET_NODE_INVALID if the node is invalid, or
+ *         RCL_RET_INVALID_ARGUMENT if any arugments are invalid, or
+ *         RCL_RET_ERROR if an unspecified error occurs.
+ */
+RCL_PUBLIC
+rcl_ret_t
+rcl_node_get_domain_id(const rcl_node_t * node, size_t * domain_id);
+
+/// Return the rmw node handle.
+/* The handle returned is a pointer to the internally held rmw handle.
+ * This function can fail, and therefore return NULL, if:
+ *   - node is NULL
+ *   - node has not been initialized (the implementation is invalid)
+ *
+ * The returned handle is only valid as long as the given node is valid.
+ *
+ * \TODO(wjwwood) should the return value of this be const?
```

Comment:
If the user accesses this handle it is likely to call some rmw specific API. Therefore I would say not const. We could always add a second API to provide a const if needs comes up.